docs/developer/design/authentication.rst

```diff
@@ -0,0 +1,160 @@
+
+Open questions
+--------------
+
+- Is support for Dex monitoring and alerting coupled with Grafana charts
+  something we should think about now?
+
+Abstract
+--------
+
+To be able to identify who a user is and to manage(Create/Edit/Delete)
+these user identities, MetalK8s needs a middleman that acts in this light
+as an Identity Provider.
+
+To accomplish this, we introduce the concept of an OpenID Connect identity
+provider which has as role identity federation in a typical MetalK8s cluster.
+
+
+Scope
+-----
+
+The scope of **Authentication** will be split into 3 main iterations as
+detailed in the implementation section of this document.
+
+Requirements
+------------
+
+User Requirements for Authentication in MetalK8s has been discussed and is
+well documented. For more information, refer to -> [link to Requirements DOC]
+
+
+Design
+------
+
+For authentication in **MetalK8s**, we are going to use Dex, Dex
+connectors/plugins and it's integrated support for other
+External IDP Integration as an Identity Provider.
+
+Design choices
+^^^^^^^^^^^^^^
+
+Dex is chosen as an Identity Provider in MetalK8s based on the above
+requirements enforced by the following reasons:
+
+* Dex's support for multiple plugins greatly increases the potential for
+  integrating with existing user management system such as Active Directory,
+  LDAP, SAML and others.
+* The ability for Dex to issue tokens, revoke tokens, refresh tokens and even
+  control the lifetime of these tokens makes **Dex** suitable since we can
+  force users to re-authenticate when need arises.
+* Dex can be seamlessly deployed in a Kubernetes cluster.
+* Dex provides access to a highly customizable UI which is a step closer to
+  good user experience which we advocate for.
+* Dex provides support for multiple backends for storage. This guarantees user
+  account information persistence after upgrades.
+
+Rejected design choices
+^^^^^^^^^^^^^^^^^^^^^^^
+
+Static passwords
+~~~~~~~~~~~~~~~~
+
+This involves adding new users by updating a static file located on every
+control-plane Node and then restarting the API server.
+
+This was rejected since there is no support for a pluggable user management
+system and it is cumbersome to operate.
+
+X.509 certificates
+~~~~~~~~~~~~~~~~~~
+
+Here, each user owns a signed certificate that is validated by the API
+server.
+
+This approach is not user-friendly meaning each certificate has to be manually
+signed. Also providing certificates for accessing the MetalK8s UI needs
+much more effort.
+
+Keycloak
+~~~~~~~~
+
+Considered as one of the heavyweight identity management systems offering a
+full blown out Identity Management.
+
+Not suitable for our use case since it entails much efforts in terms of
+deployment and management. Also since we drive for a simplistic approach to
+Identity Management, Keycloak falls short in this aspect.
+
+Keycloak is considered a future fall-back Identity Provider if the need ever
+arises from our customers.
```

Comment:
Keycloak offers everything Dex provides, and much more. We didn't choose it because of its complexity (it requires its own database and such) and also because it's overkill since we won't probably have a use-case for hosting complex user organizations (which is why we think the minimal IdP from Dex is sufficient).

It is indeed considered as fallback in case some customers require a more advanced user management **inside** MetalK8s itself (which we don't anticipate, because most of them already have such a system in place they can plug in, with either Keycloak or Dex).